File: src/HolzShots.Legacy/Drawing/Tools/UI/ScaleWindow.vb
Imports System.ComponentModel
Imports HolzShots.UI

Namespace Drawing.Tools.UI

    Friend Class ScaleWindow
        Private ReadOnly _img As Image

        <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)>
        Public Property CurrentScaleUnit As ScaleUnit = ScaleUnit.Percent
        Public ReadOnly Property WidthBoxV As Double
            Get
                Return WidthBox.DecimalValue
            End Get
        End Property
        Public ReadOnly Property HeightBoxV As Double
            Get
                Return HeightBox.DecimalValue
            End Get
        End Property

        Friend Sub New(img As Image)
            _img = img
            InitializeComponent()
        End Sub

        Private Sub ScaleWindowLoad(sender As Object, e As EventArgs) Handles MyBase.Load
            HeightBox.Text = "100"
            WidthBox.Text = "100"
        End Sub

        Private Sub PixelCheckedChanged(sender As Object, e As EventArgs) Handles Pixel.CheckedChanged
            If Pixel.Checked Then
                CurrentScaleUnit = ScaleUnit.Pixel
                WidthBox.Text = _img.Width.ToString
                HeightBox.Text = _img.Height.ToString
                UnitLabel1.Text = Localization.PixelUnit
                UnitLabel2.Text = Localization.PixelUnit
            Else
                CurrentScaleUnit = ScaleUnit.Percent
                WidthBox.Text = "100"
                HeightBox.Text = "100"
                UnitLabel1.Text = Localization.PercentUnit
                UnitLabel2.Text = Localization.PercentUnit
            End If
        End Sub

        Enum ScaleUnit
            Percent
            Pixel
        End Enum

        Private Sub HeightBoxValueChanged(sender As Object, e As EventArgs) Handles HeightBox.TextChanged
            If KeepAspectRatio.Checked AndAlso HeightBox.Focused AndAlso WidthBox.IntValue > 0 Then
                If CurrentScaleUnit = ScaleUnit.Pixel Then
                    Dim a As Double = _img.Width / _img.Height
                    Dim b As Integer = CInt(HeightBox.IntValue * a)
                    WidthBox.Text = If(b > 0, If(b > 999999, 999999, b), 1).ToString
                Else
                    If WidthBox.IntValue = HeightBox.IntValue Then Exit Sub
                    WidthBox.Text = HeightBox.IntValue.ToString
                End If
            End If
        End Sub

        Private Sub WidthBoxValueChanged(sender As Object, e As EventArgs) Handles WidthBox.TextChanged
            If KeepAspectRatio.Checked AndAlso WidthBox.Focused AndAlso HeightBox.IntValue > 0 Then
                If CurrentScaleUnit = ScaleUnit.Pixel Then
                    Dim a As Double = _img.Width / _img.Height
                    Dim b As Integer = CInt(WidthBox.IntValue / a)
                    HeightBox.Text = If(b > 0, If(b > 999999, 999999, b), 1).ToString
                Else
                    If WidthBox.IntValue = HeightBox.IntValue Then Exit Sub
                    HeightBox.Text = WidthBox.IntValue.ToString
                End If
            End If
        End Sub

        Private Sub okButton_Click(sender As Object, e As EventArgs) Handles okButton.Click
            DialogResult = DialogResult.OK
        End Sub

        Private Sub cancelButton_Click(sender As Object, e As EventArgs) Handles cnclButton.Click
            DialogResult = DialogResult.Cancel
        End Sub
    End Class
End Namespace

File: src/HolzShots.LegacyNew/Drawing/Tools/UI/ScaleWindow.cs
using System.ComponentModel;
using HolzShots.UI;

namespace HolzShots.Drawing.Tools.UI;

public partial class ScaleWindow : Form
{
    private readonly Image _img;

    [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
    public ScaleUnit CurrentScaleUnit { get; set; } = ScaleUnit.Percent;

    public double WidthBoxV => (double)WidthBox.DecimalValue;
    public double HeightBoxV => (double)HeightBox.DecimalValue;

    public ScaleWindow(Image img)
    {
        _img = img;
        InitializeComponent();
    }

    private void ScaleWindow_Load(object sender, EventArgs e)
    {
        HeightBox.Text = "100";
        WidthBox.Text = "100";
    }

    private void okButton_Click(object sender, EventArgs e)
    {
        DialogResult = DialogResult.OK;
    }
    private void cancelButton_Click(object sender, EventArgs e)
    {
        DialogResult = DialogResult.Cancel;
    }

    private void WidthBox_TextChanged(object sender, EventArgs e)
    {
        if (!KeepAspectRatio.Checked || !WidthBox.Focused || HeightBox.IntValue <= 0)
            return;

        if (CurrentScaleUnit == ScaleUnit.Pixel)
        {
            double a = (double)_img.Width / _img.Height;
            int b = (int)(WidthBox.IntValue / a);
            HeightBox.Text = (b > 0 ? (b > 999999 ? 999999 : b) : 1).ToString();
        }
        else
        {
            if (WidthBox.IntValue == HeightBox.IntValue)
                return;
            HeightBox.Text = WidthBox.IntValue.ToString();
        }
    }

    private void HeightBox_TextChanged(object sender, EventArgs e)
    {
        if (!KeepAspectRatio.Checked || !HeightBox.Focused || WidthBox.IntValue <= 0)
            return;

        if (CurrentScaleUnit == ScaleUnit.Pixel)
        {
            double a = (double)_img.Width / _img.Height;
            int b = (int)(HeightBox.IntValue * a);
            WidthBox.Text = (b > 0 ? (b > 999999 ? 999999 : b) : 1).ToString();
        }
        else
        {
            if (WidthBox.IntValue == HeightBox.IntValue)
                return;
            WidthBox.Text = HeightBox.IntValue.ToString();
        }
    }

    private void KeepAspectRatio_CheckedChanged(object sender, EventArgs e)
    {
        if (Pixel.Checked)
        {
            CurrentScaleUnit = ScaleUnit.Pixel;
            WidthBox.Text = _img.Width.ToString();
            HeightBox.Text = _img.Height.ToString();
            UnitLabel1.Text = Localization.PixelUnit;
            UnitLabel2.Text = Localization.PixelUnit;
        }
        else
        {
            CurrentScaleUnit = ScaleUnit.Percent;
            WidthBox.Text = "100";
            HeightBox.Text = "100";
            UnitLabel1.Text = Localization.PercentUnit;
            UnitLabel2.Text = Localization.PercentUnit;
        }
    }
}

public enum ScaleUnit
{
    Percent,
    Pixel
}
